packages/kepler/src/KeplerSlice.ts

```diff
@@ -207,74 +198,105 @@ export function createKeplerSlice<
             ) as Middleware;
             middlewares.push(logger);
           }
-          const dispatchWithMiddleware = (mapId: string, action: KeplerAction) => {
-            const wrapDispatch = (a: Action) => dispatch(mapId, a);
-            wrapDispatch.mapId = mapId;        
-            const middlewareAPI = {
-              getState: get,
-              dispatch: wrapDispatch,
-            };
-            // @ts-ignore
-            applyMiddleware(middlewareAPI, middlewares)(action);
+
+          const storeDispatch = (action: Action) => {
+            set((state: KeplerSliceState<PC>) => ({
+              ...state,
+              kepler: {
+                ...state.kepler,
+                map: keplerReducer(state.kepler.map, action),
+              },
+            }));
+
+            // Call onAction if it's defined
+            get().kepler.onAction?.(mapId, action);
+
+            return action;
+          };
+
+          /** Adapted from  applyMiddleware in redux */
+          let wrapDispatch: Dispatch = () => {
+            throw new Error(
+              'Dispatching while constructing your middleware is not allowed. ' +
+                'Other middleware would not be applied to this dispatch.',
+            );
+          };
+          const middlewareAPI: MiddlewareAPI = {
+            getState: get,
+            dispatch: (action, ...args): Action =>
+              wrapDispatch(action, ...args),
           };
+
+          const chain = middlewares.map((middleware) =>
+            middleware(middlewareAPI),
+          );
+          wrapDispatch = compose<Dispatch>(...chain)(storeDispatch);
           set({
             kepler: {
               ...get().kepler,
               map: keplerInitialState,
-              dispatchAction: dispatchWithMiddleware,
+              dispatchAction: (mapId, action) =>
+                wrapDispatch(wrapTo(mapId, action)),
```

Comment:
Nice, so wrapTo adds `mapId` to each action's payload, right?

Reply:
Yes, so only need to pass `action` to `wrapDispatch`, because the middleware doesn't take `mapId, action` as input. We need to pass regular `dispatch` to middleware for chaining